### PR TITLE
Create cluster.cr

### PR DESCRIPTION
**Pull request by adrien-thierry**
Add a Cluster-like capacity to Crystal

### CODE REVIEW

File: src/cluster.cr [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+    def self.isMaster
+        (ENV["FORKED"] ||= "0") == "0"
+    end
+
+    def self.isSlave
+        (ENV["FORKED"] ||= "0") == "1"
+    end
+
+    def self.getEnv (env : String)
```
Comment (RX14) [inline src/cluster.cr]: This method is unneccesary

File: src/cluster.cr [hunk 2]
```diff
@@ -0,0 +1,20 @@
+class Cluster
+
+    def self.fork (env : Hash)
+        env["FORKED"] = "1"
```
Comment (RX14) [inline src/cluster.cr]: I would namespace this env var more.

**Comment by RX14** [comment]
I would like to see a forking cluster helper in the standard library, however I'm a little dubious on the API. At the very least needs a format and docs.

**Comment by adrien-thierry** [comment]
Why not take inspiration from the nodejs Cluster api? https://nodejs.org/docs/latest/api/cluster.html

File: src/cluster.cr [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        return Process.fork { Process.run(PROGRAM_NAME, nil, env, true, false, true, true, true, nil ) }
+    end
+
+    def self.isMaster
```
Comment (schovi) [inline src/cluster.cr]: `self.master?` is more Crystal/Ruby semantic. Same for `self.slave?`

Comment (adrien-thierry) [inline src/cluster.cr]: When I mention the NodeJS Cluster API, I speak about the API, not the semantic

Comment (Sija) [inline src/cluster.cr]: NodeJS API adheres to JS conventions, not to Ruby ones. It's a convention to use `isFoo` in JS, whereas in Ruby you'd rather write `foo?`.

File: src/cluster.cr [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+        (ENV["FORKED"] ||= "0") == "0"
```
Comment (Sija) [inline src/cluster.cr]: You shouldn't set value in checker method. Proper way would be `(ENV["FORKED"]? || "0") == "0"`.

Comment (adrien-thierry) [inline src/cluster.cr]: Done

File: src/cluster.cr (hunk 3 above)
Comment (adrien-thierry) [inline src/cluster.cr]: I understand, changes done

File: src/cluster.cr [hunk 5]
```diff
@@ -0,0 +1,16 @@
+class Cluster
+
+    def self.fork (env : Hash)
+        env["FORKED"] = "1"
+        return Process.fork { Process.run(PROGRAM_NAME, nil, env, true, false, true, true, true, nil ) }
+    end
+
+    def self.master
```
Comment (Sija) [inline src/cluster.cr]: `def self.master?`

**Comment by Sija** [comment]
@adrien-thierry NodeJS cluster comes with more than just wrapping a child process with ENV parameter. Quoting from the docs:

> A single instance of Node.js runs in a single thread. To take advantage of multi-core systems the user will sometimes want to launch a cluster of Node.js processes to handle the load.
>
> The cluster module allows you to easily create child processes that all share server ports.

> The worker processes are spawned using the child_process.fork() method, so that they can communicate with the parent via IPC and pass server handles back and forth.

That said, having similar capabilities in Crystal would be great but IMHO this PR in its current doesn't add much value to stdlib.

File: src/cluster.cr [hunk 6]
```diff
@@ -0,0 +1,16 @@
+class Cluster
+
+    def self.fork (env : Hash)
+        env["FORKED"] = "1"
+        return Process.fork { Process.run(PROGRAM_NAME, nil, env, true, false, true, true, true, nil ) }
```
Comment (Sija) [inline src/cluster.cr]: redundant `return`

**Comment by adrien-thierry** [comment]
@Sija With SO_REUSEPORT support, it's a goot start to add easy multithreading-like capabilities to http server. As is, it lacks events (close etc) and IPC

**Comment by bcardiff** [comment]
It's better to see something like this evolve as a separate shard first.

I would suggest also to take a look at https://github.com/ysbaddaden/panzer .

Thanks!

**Pull request closed without merging**